File: core/src/main/kotlin/org/anvilpowered/anvil/core/config/ConfigurateRegistry.kt
/*
 *   Anvil - AnvilPowered.org
 *   Copyright (C) 2019-2026 Contributors
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.core.config

import org.anvilpowered.anvil.core.config.ConfigurateRegistry.Factory.DiscoveryClosure
import org.apache.logging.log4j.Logger
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.serialize.TypeSerializerCollection
import java.nio.file.Path
import kotlin.io.path.createDirectory
import kotlin.io.path.extension
import kotlin.io.path.listDirectoryEntries
import kotlin.io.path.notExists

class ConfigurateRegistry(
  private val rootNode: ConfigurationNode,
  private val delegate: Registry? = null,
) : Registry {
  override fun <T : Any> getDefault(key: Key<T>): T = delegate?.getDefault(key) ?: key.fallback
  override fun <E : Any> getDefault(key: ListKey<E>, index: Int): E {
    return delegate?.getDefault(key, index)
      ?: key.fallback.getOrNull(index)
      ?: throw NoSuchElementException("No default value for key ${key.name} at index $index")
  }

  override fun <K : Any, V : Any> getDefault(key: MapKey<K, V>, mapKey: K): V {
    return delegate?.getDefault(key, mapKey)
      ?: key.fallback[mapKey]
      ?: throw NoSuchElementException("No default value for key ${key.name} at key $mapKey")
  }

  override fun <T : Any> getStrict(key: SimpleKey<T>): T? = rootNode.node(key.configNodePath)[key.type]
  override fun <E : Any> getStrict(key: ListKey<E>): List<E>? = rootNode.node(key.configNodePath)[key.type]
  override fun <E : Any> getStrict(key: ListKey<E>, index: Int): E? = getStrict(key)?.let { return it[index] }
  override fun <K : Any, V : Any> getStrict(key: MapKey<K, V>): Map<K, V>? = rootNode.node(key.configNodePath)[key.type]
  override fun <K : Any, V : Any> getStrict(key: MapKey<K, V>, mapKey: K): V? = getStrict(key)?.let { return it[mapKey] }

  companion object Factory {

    data class DiscoverResult internal constructor(
      val registry: Registry,
      val path: Path,
      val type: ConfigurateFileType<*>,
    )

    fun discover(
      basePath: Path,
      logger: Logger,
      serializers: TypeSerializerCollection = TypeSerializerCollection.defaults(),
      delegate: Registry? = null,
    ): DiscoverResult? {
      if (basePath.notExists()) {
        basePath.createDirectory()
      }

      val configFiles = basePath.listDirectoryEntries()
        .map { it to ConfigurateFileType.fromName(it.extension) }
        .mapNotNull { (path, type) -> type?.let { path to it } }
        .toList()

      if (configFiles.isEmpty()) {
        return null
      } else if (configFiles.size >= 2) {
        throw IllegalStateException(
          "Detected multiple configuration files for plugin ${basePath.fileName}: ${configFiles.map { it.first }}. " +
            "Please make sure there is only one configuration file per plugin",
        )
      }

      val (path, type) = configFiles.single()
      return DiscoverResult(ConfigurateRegistry(type.createBuilder(serializers).path(path).build().load(), delegate), path, type)
    }

    fun createDiscoveryClosure(
      basePath: Path,
      logger: Logger,
      serializers: TypeSerializerCollection = TypeSerializerCollection.defaults(),
      delegate: Registry? = null,
    ) = DiscoveryClosure { discover(basePath, logger, serializers, delegate) }

    fun interface DiscoveryClosure {
      fun discover(): DiscoverResult?
    }
  }
}

val Key<*>.configNodePath: List<String>
  get() = name.split('_').map { it.lowercase() }

File: core/src/main/scala/org/anvilpowered/anvil/core/config/ConfigurateRegistry.scala
/*
 *   Anvil - AnvilPowered.org
 *   Copyright (C) 2019-2026 Contributors
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.core.config

import cats.data.{EitherT, OptionT}
import cats.effect.kernel.Resource
import cats.effect.{Async, Concurrent, IO}
import cats.syntax.all.*
import fs2.io.file.{Files, Path}
import org.anvilpowered.anvil.core.config.ConfigurateRegistry.getConfigNodePath
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.serialize.TypeSerializerCollection

import scala.jdk.CollectionConverters.IterableHasAsScala

class ConfigurateRegistry(
    private val rootNode: ConfigurationNode,
    private val delegate: Option[Registry],
) extends Registry {
  override def getDefault[T](key: Key[T]): T = delegate.map(_.getDefault(key)).getOrElse(key.fallback)
  override def get[T](key: Key[T]): Option[T] = Option(rootNode.node(key.getConfigNodePath).get(key.typeToken))
}

object ConfigurateRegistry {
  case class DiscoverResult(registry: Registry, path: Path, fileType: ConfigurateFileType[?])

  def discover[F[_]](
      basePath: Path,
      serializers: TypeSerializerCollection = TypeSerializerCollection.defaults(),
      delegate: Option[Registry] = None,
  )(using F: Async[F]): EitherT[F, String, DiscoverResult] =
    for {
      configFiles <- EitherT.liftF(
        Files.forAsync[F]
          .list(basePath)
          .mapFilter(path => ConfigurateFileType.fromName(path.extName).map(path -> _))
          .compile
          .toList,
      )
      (path, fileType) <- EitherT.cond(
        configFiles.size == 1,
        configFiles.head,
        s"Detected multiple configuration files for plugin ${basePath.fileName}: ${configFiles.map(_._1)}. " +
          "Please make sure there is only one configuration file per plugin",
      )
    } yield {
      val rootNode = fileType.createBuilder(serializers).path(path.toNioPath).build().load()
      val registry = ConfigurateRegistry(rootNode, delegate)
      DiscoverResult(registry, path, fileType)
    }

  extension (key: Key[?]) {
    def getConfigNodePath: List[String] = key.name.split('_').map(_.toLowerCase).toList
  }
}
